File: LeetCode_594/LongestHarmoniousSubsequence.java
package LeetCode_594;


import java.util.Arrays;
import java.util.Scanner;

class Solution {
    public int findLHS(int[] nums) {
        Arrays.sort(nums);
        int prevCount = 0, currCount = 1, result = 0;
        int prevNum = nums[0];

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] == nums[i - 1]) {
                currCount++;
            } else {
                if (nums[i - 1] - 1 == prevNum) {
                    result = Math.max(result, prevCount + currCount);
                }
                prevCount = currCount;
                prevNum = nums[i - 1];
                currCount = 1;
            }
        }

        if (nums[nums.length - 1] - 1 == prevNum) {
            result = Math.max(result, prevCount + currCount);
        }

        return result;
    }
}

public class LongestHarmoniousSubsequence {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();
        int[] nums = new int[n];

        for (int i = 0; i < n; i++) {
            nums[i] = scanner.nextInt();
        }

        Solution sol = new Solution();
        System.out.println(sol.findLHS(nums));

        scanner.close();
    }
}

File: LeetCode_594/LongestHarmoniousSubsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */

function max(x, y) {
    return x > y ? x : y;
}

var findLHS = function(nums) {
    nums.sort((a, b) => a - b);
    let numsSize = nums.length;
    let prev_cnt = 0, cur_cnt = 1, res = 0;
    let prev_num = 0;

    for (let i = 0; i + 1 < numsSize; i++) {
        if (nums[i] == nums[i + 1]) {
            cur_cnt += 1;
        } else {
            if (prev_cnt) {
                if (prev_num + 1 == nums[i]) {
                    res = max(res, prev_cnt + cur_cnt);
                }
            }
            prev_cnt = cur_cnt;
            prev_num = nums[i];
            cur_cnt = 1;
        }
    }

    if (prev_cnt) {
        if (prev_num + 1 == nums[numsSize - 1]) {
            res = max(res, prev_cnt + cur_cnt);
        }
    }
    return res;
};

const nums = [1,3,2,2,5,2,3,7];
console.log(findLHS(nums));